docs/source/reference/configuration.rst
.. _configuration:

Configuration
=============

Once you have installed wis2box, it is time to set up the configuration. The wis2box setup is based on
a simple configuration that can be adjusted depending the user's needs and deployment environment.

Environment variables
---------------------

wis2box configuration is driven primarily by a set of environment variables. The runtime
configuration is defined in the `Env`_ format in a plain text file named ``wis2box.env``. 
An example is provided in ``wis2box.env.example``.

You can either copy the example-file to ``wis2box.env`` and adjust the values to your needs or run the following command
to create a new ``wis2box.env`` file by answering a few questions on the command line:

.. code-block:: bash

    python3 wis2box-create-config.py

For further / specialized configuration, see the sections below.

``WIS2BOX_HOST_DATADIR``
^^^^^^^^^^^^^^^^^^^^^^^^

The value of ``WIS2BOX_HOST_DATADIR`` maps the wis2box data directory from the host system to the containers.


Sections
--------

.. note::

   A reference configuration can always be found in the wis2box `GitHub`_ repository. The :ref:`quickstart`
   uses a variant of ``wis2box.env.example`` with mappings to the test data, as an example.

wis2box environment variables can be categorized via the following core sections:

- **Contact/identification**: Contact and identification information
- **Storage**: MinIO configuration
- **API**: API configuration for provisioning the OGC API capabilities
- **Logging**: logging configuaration for wis2box
- **Pub/Sub**: Pub/Sub options
- **Other**: other miscellaneous options

.. note::

    Configuration directives and reference are described below via annotated examples. Changes in configuration
    require a restart of wis2box to take effect. See the :ref:`administration` section for information on
    managing wis2box.

Contact / identification
^^^^^^^^^^^^^^^^^^^^^^^^

wis2box provides the ability to customize the contact and identification information for a given installation.  This
is displayed on the public landing page of the API.  If any of these items are not set, default contact/identification
information is set.

.. code-block:: bash

    # UI customizations
    WIS2BOX_ICON=https://example.org/icon
    WIS2BOX_LOGO=https://example.org/logo

    # service identification
    WIS2BOX_IDENTIFICATION_TITLE=title
    WIS2BOX_IDENTIFICATION_DESCRIPTION=description
    WIS2BOX_IDENTIFICATION_TERMS_OF_SERVICE=CC-BY 4.0
    WIS2BOX_IDENTIFICATION_URL=https://example.org/tos
    # data license information
    WIS2BOX_LICENSE_NAME=CC-BY 4.0
    WIS2BOX_LICENSE_URL=https://creativecommons.org/licenses/by/4.0
    # service provider information
    WIS2BOX_PROVIDER_NAME=Firstname Lastname
    WIS2BOX_PROVIDER_URL=https://example.org/provider
    # contact information
    WIS2BOX_CONTACT_NAME=Firstname Lastname
    WIS2BOX_CONTACT_POSITION=staff
    WIS2BOX_CONTACT_ADDRESS=123 Main street
    WIS2BOX_CONTACT_CITY=city
    # NOTE: if your postal code is only numeric, prefix with the text "code " to ensure it is interpreted as a string
    WIS2BOX_CONTACT_POSTALCODE=postal code
    WIS2BOX_CONTACT_COUNTRY=country
    WIS2BOX_CONTACT_EMAIL=you@example.org
    WIS2BOX_CONTACT_URL=https://example.org/contact
    WIS2BOX_CONTACT_INSTRUCTIONS=contact instructions
    WIS2BOX_CONTACT_ROLE=host


Storage
^^^^^^^

wis2box currently supports S3 compatible storage (e.g. MinIO, Amazon S3). Additional storage types are planned for
future releases.

The following environment variables can be used to configure `WIS2BOX_STORAGE`.

.. note::

   When using wis2box in production and using the default MinIO-container, please specify a unique ``WIS2BOX_STORAGE_PASSWORD``

.. code-block:: bash

    WIS2BOX_STORAGE_TYPE=S3
    WIS2BOX_STORAGE_SOURCE=http://minio:9000
    WIS2BOX_STORAGE_USERNAME=wis2box  # username for the storage-layer
    WIS2BOX_STORAGE_PASSWORD=minio123  # password for the storage-layer
    WIS2BOX_STORAGE_INCOMING=wis2box-incoming  # name of the storage-bucket/folder for incoming files
    WIS2BOX_STORAGE_PUBLIC=wis2box-public  # name of the storage-bucket/folder for public files
    WIS2BOX_STORAGE_DATA_RETENTION_DAYS=7  # number of days to keep files in incoming and public
    WIS2BOX_STORAGE_API_RETENTION_DAYS=7  # number of days to keep files in API backend


MinIO
^^^^^

wis2box uses MinIO as the default S3 storage capability.

When overriding the default storage environment variables, please redefine the ``MINIO*`` environment variables to match
your configuration.

.. code-block:: bash

    MINIO_ROOT_USER=${WIS2BOX_STORAGE_USERNAME}
    MINIO_ROOT_PASSWORD=${WIS2BOX_STORAGE_PASSWORD}
    MINIO_NOTIFY_MQTT_USERNAME_WIS2BOX=${WIS2BOX_BROKER_USERNAME}
    MINIO_NOTIFY_MQTT_PASSWORD_WIS2BOX=${WIS2BOX_BROKER_PASSWORD}
    MINIO_NOTIFY_MQTT_BROKER_WIS2BOX=tcp://${WIS2BOX_BROKER_HOST}:${WIS2BOX_BROKER_PORT}


API
^^^

API configurations drive control of the OGC API setup.

.. code-block:: bash

    WIS2BOX_API_TYPE=pygeoapi  # server type
    WIS2BOX_API_URL=http://localhost/pygeoapi  # public landing page endpoint
    WIS2BOX_API_BACKEND_TYPE=Elasticsearch  # backend provider type
    WIS2BOX_API_BACKEND_URL=http://elasticsearch:9200  # internal backend connection URL
    WIS2BOX_DOCKER_API_URL=http://wis2box-api:80/oapi  # container name of API container (for internal communications/workflow)

Logging
^^^^^^^

The logging directives control logging level/severity and output.

.. code-block:: bash

    WIS2BOX_LOGGING_LOGLEVEL=ERROR  # the logging level (see https://docs.python.org/3/library/logging.html#logging-levels)
    WIS2BOX_LOGGING_LOGFILE=stdout  # the full file path to the logfile or ``stdout`` to display on console

Pub/Sub
^^^^^^^

Pub/Sub configuration provides connectivity information for the Pub/Sub broker.

.. code-block:: bash

    WIS2BOX_BROKER_HOST=mosquitto  # the hostname of the internal broker
    WIS2BOX_BROKER_PORT=1883  # the port of the internal broker
    WIS2BOX_BROKER_USERNAME=wis2box  # the username of the internal broker
    WIS2BOX_BROKER_PASSWORD=wis2box  # the password of the internal broker
    WIS2BOX_BROKER_PUBLIC=mqtt://foo:bar@localhost:1883  # RFC 1738 URL of public broker endpoint
    WIS2BOX_BROKER_QUEUE_MAX=1000  # maximum number of messages to hold in the queue per client

.. note::

   ``WIS2BOX_BROKER_QUEUE_MAX`` should be configured according to the setup of wis2box, relative to the number
   of expected observations per day.  See :ref:`getting-started` for more information on system requirements.


Note that the ``WIS2BOX_BROKER_PUBLIC`` URL can be used to publish WIS2 notifications to an external broker. By default, the internal broker is used.

Apart from the wis2box internal user defined by the ``WIS2BOX_BROKER_USERNAME`` and ``WIS2BOX_BROKER_PASSWORD`` environment variables, the wis2box broker will also include the user ``everyone`` with password ``everyone``.

The ``everyone`` user has **read-only** access to the ``origin/#`` topic and can be used to allow the WIS2 Global Broker to subscribe to the wis2box.

To add additional users to the wis2box broker, login to the mosquitto container with the following command:

.. code-block:: bash

    docker exec -it mosquitto /bin/sh

Then, to add a new user, use the following command:

.. code-block:: bash

    mosquitto_passwd -b /mosquitto/config/password.txt <username> <password>

To add or change access rights for mosquitto users, you can edit the file ``/mosquitto/config/acl.conf`` from inside the mosquitto container using the ``vi`` command:

.. code-block:: bash

    vi /mosquitto/config/acl.conf

See the mosquitto documentation for more information on the ACL configuration file.

Restart the mosquitto container for the changes to take effect with the command:

.. code-block:: bash

    docker restart mosquitto

Web application
^^^^^^^^^^^^^^^

Web application configuration provides the ability to customize web components.
All of the below directives are optional.

.. code-block:: bash

    WIS2BOX_BASEMAP_URL="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"  # URL of map tile server to use
    WIS2BOX_BASEMAP_ATTRIBUTION="<a href="https://osm.org/copyright">OpenStreetMap</a> contributors"  # attribution of map tile server
    WIS2BOX_UI_CLUSTER=False # default setting of the cluster toggle
    WIS2BOX_UI_LANG="en" # default language, one of: ar, en, es, fr, ru, zh
    WIS2BOX_UI_LOGO=http://example.com/logo.png # use custom organization logo in the UI
    WIS2BOX_UI_BANNER_COLOR="#014e9e" # use custom background color for header and footer 
    WIS2BOX_UI_DISABLE_SEPARATOR_IMAGE=false # boolean to disable WMO separator

.. note::
    ``WIS2BOX_UI_LOGO`` requires a full URL to the image file.

    If you want to use a local image you can upload your image in the "wis2box-public" bucket of the MinIO storage and use the URL of the image in the configuration.

    Note that the web proxy exposes the "wis2box-public" bucket as the ``/data/`` endpoint.
    If your wis2box uses the URL ``https://wis2box.example.com`` and the uploaded image is named ``logo.png``, 
    you would set ``WIS2BOX_UI_LOGO=https://wis2box.example.com/data/logo.png``.

Other
^^^^^

Additional directives provide various configuration options for the deployment of wis2box.

.. code-block:: bash

    WIS2BOX_URL=http://localhost/  # public wis2box url
    WIS2BOX_AUTH_STORE=http://wis2box-auth # wis2box auth service location


.. note::

   To access internal containers, URL configurations should point to the named containers as specified in ``docker-compose.yml``.


A full configuration example can be found below:

.. literalinclude:: ../../../wis2box.env.example
   :language: bash

Docker Compose
--------------

The Docker Compose set up is driven from the resulting ``wis2box.env`` file created. For advanced cases and/or power users,
updates can also be made to ``docker-compose.yml`` or ``docker-compose.override.yml`` (for changes to ports).

.. _`Env`: https://en.wikipedia.org/wiki/Env
.. _`GitHub`: https://github.com/World-Meteorological-Organization/wis2box/blob/main/default.env


wis2box-management/wis2box/env.py
###############################################################################
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
###############################################################################

import click
import logging
import os
from pathlib import Path

from wis2box import cli_helpers
from wis2box.log import setup_logger
from wis2box.plugin import load_plugin
from wis2box.plugin import PLUGINS
from wis2box.util import remove_auth_from_url

LOGGER = logging.getLogger(__name__)

try:
    DATADIR = Path(os.environ.get('WIS2BOX_DATADIR'))
except (OSError, TypeError):
    msg = 'Configuration filepaths do not exist!'
    LOGGER.error(msg)
    raise EnvironmentError(msg)

API_TYPE = os.environ.get('WIS2BOX_API_TYPE', 'pygeoapi')
API_URL = os.environ.get('WIS2BOX_API_URL', 'http://localhost/oapi')
API_BACKEND_TYPE = os.environ.get('WIS2BOX_API_BACKEND_TYPE', 'Elasticsearch')
API_BACKEND_URL = os.environ.get('WIS2BOX_API_BACKEND_URL', 'http://elasticsearch:9200').rstrip('/') # noqa
DOCKER_API_URL = os.environ.get('WIS2BOX_DOCKER_API_URL', 'http://wis2box-api:80') # noqa
AUTH_URL = os.environ.get('WIS2BOX_AUTH_URL', 'http://wis2box-auth')
URL = os.environ.get('WIS2BOX_URL', 'http://localhost')

BROKER_USERNAME = os.environ.get('WIS2BOX_BROKER_USERNAME', 'wis2box')
BROKER_PASSWORD = os.environ.get('WIS2BOX_BROKER_PASSWORD', 'wis2box')
BROKER_HOST = os.environ.get('WIS2BOX_BROKER_HOST', 'mosquitto')
BROKER_PORT = os.environ.get('WIS2BOX_BROKER_PORT', 1883)
DOCKER_BROKER = f'mqtt://{BROKER_USERNAME}:{BROKER_PASSWORD}@{BROKER_HOST}:{BROKER_PORT}' # noqa
BROKER_PUBLIC = os.environ.get('WIS2BOX_BROKER_PUBLIC', DOCKER_BROKER)

STORAGE_TYPE = os.environ.get('WIS2BOX_STORAGE_TYPE', 'S3')
STORAGE_SOURCE = os.environ.get('WIS2BOX_STORAGE_SOURCE', 'http://minio:9000')
STORAGE_USERNAME = os.environ.get('WIS2BOX_STORAGE_USERNAME', 'wis2box')
STORAGE_PASSWORD = os.environ.get('WIS2BOX_STORAGE_PASSWORD', 'minio123')
STORAGE_INCOMING = os.environ.get('WIS2BOX_STORAGE_INCOMING', 'wis2box-incoming') # noqa
STORAGE_PUBLIC = os.environ.get('WIS2BOX_STORAGE_PUBLIC', 'wis2box-public')

try:
    STORAGE_DATA_RETENTION_DAYS = int(os.environ.get('WIS2BOX_STORAGE_DATA_RETENTION_DAYS')) # noqa
except TypeError:
    STORAGE_DATA_RETENTION_DAYS = None

try:
    STORAGE_API_RETENTION_DAYS = int(os.environ.get('WIS2BOX_STORAGE_API_RETENTION_DAYS', 100)) # noqa
except TypeError:
    STORAGE_API_RETENTION_DAYS = None

LOGLEVEL = os.environ.get('WIS2BOX_LOGGING_LOGLEVEL', 'ERROR')
LOGFILE = os.environ.get('WIS2BOX_LOGGING_LOGFILE', 'stdout')

missing_environment_variables = []

required_environment_variables = [
    DATADIR,
    DOCKER_API_URL,
    API_TYPE,
    BROKER_HOST,
    BROKER_PORT,
    BROKER_USERNAME,
    BROKER_PASSWORD,
    BROKER_PUBLIC,
    URL,
    STORAGE_TYPE,
    STORAGE_SOURCE,
    STORAGE_USERNAME,
    STORAGE_PASSWORD,
    STORAGE_INCOMING,
    STORAGE_PUBLIC
]

for rev in required_environment_variables:
    if rev is None:
        envvar_name = [k for k, v in locals().items() if v is rev][0]
        LOGGER.warning(f'Missing environment variable {envvar_name}')
        missing_environment_variables.append(envvar_name)

if missing_environment_variables:
    msg = f'Environment variables not set! = {missing_environment_variables}'
    LOGGER.error(msg)
    raise EnvironmentError(msg)

if '@' in BROKER_PASSWORD:
    msg = 'WIS2BOX_BROKER_PASSWORD must not contain "@" character'
    LOGGER.error(msg)
    raise EnvironmentError(msg)

if '@' in STORAGE_PASSWORD:
    msg = 'WIS2BOX_STORAGE_PASSWORD must not contain "@" character'
    LOGGER.error(msg)
    raise EnvironmentError(msg)


@click.group()
def environment():
    """Environment management"""
    pass


@click.command()
@click.pass_context
@cli_helpers.OPTION_VERBOSITY
def test(ctx, verbosity):
    """Tests the environment is set up correctly"""

    click.echo(f'Setting up logging (loglevel={LOGLEVEL}, logfile={LOGFILE})')
    setup_logger(LOGLEVEL, LOGFILE)

    click.echo('Testing BROKER_PUBLIC')
    # load plugin for plugin-broker
    defs = {
        'codepath': PLUGINS['pubsub']['mqtt']['plugin'],
        'url': BROKER_PUBLIC,
        'client_type': 'publisher'
    }
    broker = load_plugin('pubsub', defs)

    try:
        result = broker.test(topic='origin/a/wis2/test', message='wis2box pub test at startup') # noqa
    except Exception as err:
        LOGGER.error(err)
        raise EnvironmentError(err)

    if result:
        click.echo('Broker test successful')
    else:
        LOGGER.error('Could not connect to broker defined by WIS2BOX_BROKER_PUBLIC') # noqa
        click.echo('Broker test failed')
        exit(1)


@click.command()
@click.pass_context
@cli_helpers.OPTION_VERBOSITY
def create(ctx, verbosity):
    """Creates baseline data/metadata directory structure"""

    click.echo(f'Setting up logging (loglevel={LOGLEVEL}, logfile={LOGFILE})')
    setup_logger(LOGLEVEL, LOGFILE)

    click.echo('Setting up storage')
    storage_defs = {
        'storage_type': STORAGE_TYPE,
        'source': STORAGE_SOURCE,
        'auth': {'username': STORAGE_USERNAME, 'password': STORAGE_PASSWORD},
        'codepath': PLUGINS['storage'][STORAGE_TYPE]['plugin']
    }

    storages = {
        STORAGE_INCOMING: 'private',
        STORAGE_PUBLIC: 'readonly'
    }
    for key, value in storages.items():
        storage_defs['name'] = key
        storage_defs['policy'] = value
        storage = load_plugin('storage', storage_defs)
        storage.setup()

    # TODO: abstract into wis2box.storage.fs.FileSystemStorage
    click.echo(f'Creating baseline directory structure in {DATADIR}')
    DATADIR.mkdir(parents=True, exist_ok=True)
    (DATADIR / 'metadata' / 'discovery').mkdir(parents=True, exist_ok=True)
    (DATADIR / 'metadata' / 'station').mkdir(parents=True, exist_ok=True)


@click.command()
@click.pass_context
@cli_helpers.OPTION_VERBOSITY
def show(ctx, verbosity):
    """Displays wis2box environment variables"""

    for key, value in os.environ.items():
        if key.startswith('WIS2BOX'):
            if 'PASSWORD' in key:
                value2 = '*' * len(value)
            elif key == 'WIS2BOX_BROKER_PUBLIC':
                value2 = remove_auth_from_url(value)
            else:
                value2 = value

            click.echo(f'{key} => {value2}')


environment.add_command(create)
environment.add_command(show)
environment.add_command(test)
